### ISSUE TITLE
unused method tear offs

### ISSUE THREAD

**Issue by jmesserly** [dart-lang/linter#269]
This came up in real code, in the middle of a cascade (`list..clear..add(123)`).

For example:

``` dart
int f() => 1;
class C {
  int m() => 2;
}

main() {
  f;
  new C().m;
  var c = new C()..m;
}
```

The only analyzer message from that is the "unused c". I glanced through the hints and didn't see any about this specifically.

In general, any non-side-effecting expression used as a statement might be worth reporting (e.g. `'hello';`, but I think there's especially value in reporting it for method tear offs.

p.s. I wasn't sure if this should be a lint or an Analyzer hint, feels like it could go either way, maybe a lint is easier to start with?

**Comment by pq** [comment]
cc @bwilkerson on the hint vs. lint question

**Comment by bwilkerson** [comment]
My answer usually comes back to the questions of how many false positives there will be and whether this is really a bug in the code or just a style issue. In Dart, it's relatively hard to know when an expression cannot have a side effect, but tear-offs and non-user-definable operators are clear examples. I'd make it a hint because I think its clearly a bug and it shouldn't have any false positives.

**Comment by jmesserly** [comment]
yeah, that makes sense, thanks!

BTW I forgot to mention, this was in internal code (readability review), I can send you a pointer to the CL if it helps see it in context.

**Comment by a14n** [comment]
With #967 all reported examples are now correctly lint by `unnecessary_statements`. This issue can be closed.
